model/channel_shared.go

```diff
@@ -0,0 +1,166 @@
+	TeamId           string `json:"team_id"`
+	Home             bool   `json:"home"`
+	ReadOnly         bool   `json:"readonly"`
+	ShareName        string `json:"share_name"`
+	ShareDisplayName string `json:"share_displayname"`
+	SharePurpose     string `json:"share_purpose"`
+	ShareHeader      string `json:"share_header"`
+	CreatorId        string `json:"creator_id"`
+	CreateAt         int64  `json:"create_at"`
+	UpdateAt         int64  `json:"update_at"`
+	RemoteClusterId  string `json:"remote_cluster_id,omitempty"` // if not "home"
+	Token            string `json:"token,omitempty"`             // if not "home"
+}
+
+func (sc *SharedChannel) ToJson() string {
+	b, _ := json.Marshal(sc)
+	return string(b)
+}
+
+func SharedChannelFromJson(data io.Reader) (*SharedChannel, error) {
+	var sc *SharedChannel
+	err := json.NewDecoder(data).Decode(&sc)
+	return sc, err
+}
+
+func (sc *SharedChannel) IsValid() *AppError {
+	if !IsValidId(sc.ChannelId) {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.id.app_error", nil, "ChannelId="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if !IsValidId(sc.TeamId) {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.id.app_error", nil, "TeamId="+sc.TeamId, http.StatusBadRequest)
+	}
+
+	if sc.CreateAt == 0 {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.create_at.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if sc.UpdateAt == 0 {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.update_at.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if utf8.RuneCountInString(sc.ShareDisplayName) > CHANNEL_DISPLAY_NAME_MAX_RUNES {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.display_name.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if !IsValidChannelIdentifier(sc.ShareName) {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.2_or_more.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if utf8.RuneCountInString(sc.ShareHeader) > CHANNEL_HEADER_MAX_RUNES {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.header.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if utf8.RuneCountInString(sc.SharePurpose) > CHANNEL_PURPOSE_MAX_RUNES {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.purpose.app_error", nil, "id="+sc.ChannelId, http.StatusBadRequest)
+	}
+
+	if !IsValidId(sc.CreatorId) {
+		return NewAppError("Channel.IsValid", "model.channel.is_valid.creator_id.app_error", nil, "CreatorId="+sc.CreatorId, http.StatusBadRequest)
+	}
+
+	if !sc.Home {
+		if !IsValidId(sc.RemoteClusterId) {
+			return NewAppError("Channel.IsValid", "model.channel.is_valid.id.app_error", nil, "RemoteClusterId="+sc.RemoteClusterId, http.StatusBadRequest)
+		}
+		if !IsValidId(sc.Token) {
+			return NewAppError("Channel.IsValid", "model.channel.is_valid.id.app_error", nil, "Token", http.StatusBadRequest)
```

Comment:
maybe log the invalid token value

Reply:
What can't log tokens, even invalid ones.  Tokens are treated like passwords.